scanpy/experimental/pp/_normalization.py

```diff
@@ -0,0 +1,243 @@
+    `theta` shared across genes. By default, residuals are clipped to sqrt(n),
+    overdispersion `theta=100` is used, and PCA is run with 50 components.
+
+    Operates on the subset of highly variable genes in `adata.var['highly_variable']`
+    by default. Expects raw count input.
+
+
+    Params
+    ------
+    adata
+        The annotated data matrix of shape `n_obs` × `n_vars`.
+        Rows correspond to cells and columns to genes.
+    theta
+        The negative binomial overdispersion parameter theta for Pearson residuals.
+        Higher values correspond to less overdispersion (var = mean + mean^2/theta),
+        and `theta=np.Inf` corresponds to a Poisson model.
+    clip
+        Determines if and how residuals are clipped:
+
+            * If `None`, residuals are clipped to the interval [-sqrt(n), sqrt(n)], \
+            where n is the number of cells in the dataset (default behavior).
+            * If any scalar c, residuals are clipped to the interval [-c, c]. Set \
+            `clip=np.Inf` for no clipping.
+
+    n_comps_pca
+        Number of principal components to compute.
+    random_state_pca
+        Change to use different initial states for the optimization.
+    kwargs_pca
+        Dictionary of further keyword arguments passed on to `scanpy.pp.pca()`.
+    use_highly_variable
+        Whether to use the gene selection in `adata.var['highly_variable']` to subset
+        the data before normalizing (default) or proceed on the full dataset.
+    check_values
+        Check if counts in selected layer are integers. A Warning is returned if set to
+        True.
+    inplace
+        Whether to place results in `adata` or return them.
+
+
+    Returns
+    -------
+    If `inplace=False`, returns the Pearson residual-based PCA results
+    (`adata_pca`).
+    If `inplace=True`, updates `adata` with the following fields:
+
+    `.uns['pearson_residuals_normalization']['pearson_residuals_df']`
+         The hvg-subset, normalized by Pearson residuals
+    `.uns['pearson_residuals_normalization']['theta']`
+         The used value of the overdisperion parameter theta
+    `.uns['pearson_residuals_normalization']['clip']`
+         The used value of the clipping parameter
+
+    `.obsm['X_pca']`
+        PCA representation of data after gene selection (if applicable) and Pearson
+        residual normalization.
+    `.uns['pca']['PCs']`
+         The principal components containing the loadings.
+    `.uns['pca']['variance_ratio']`
+         Ratio of explained variance.
+    `.uns['pca']['variance']`
+         Explained variance, equivalent to the eigenvalues of the covariance matrix.
+
+    """
+
+    if use_highly_variable and 'highly_variable' in adata.var_keys():
+        # TODO: are these copies needed?
+        adata_pca = adata[:, adata.var['highly_variable']].copy()
+    else:
+        # TODO: are these copies needed?
+        adata_pca = adata.copy()
```

Comment:
Two things:

* If highly variable is explicitly passed as True, but `"highly_variable"` isn't it `.var`, I think this should probably error. This can be rectified by making the argument default to `None`, then become `True` if `"highly_variable"` is present
* This currently copies all of the data in the passed object, while all it really needs is the matrix to compute transform and compute PCA on

```suggestion
    if use_highly_variable is None and 'highly_variable' in adata.var_keys():
        use_highly_variable = True

    if use_highly_variable:
        adata_sub = adata[:, adata.var['highly_variable']].copy()
        adata_pca = AnnData(adata_sub.X.copy(), obs=adata_sub.obs[[]], var=adata_sub.var[[]])
    else:
        adata_pca = AnnData(adata.X.copy(), obs=adata.obs[[]], var=adata.var[[]])
```

Reply:
took you suggested code and added a check for the case you said should error :heavy_check_mark: